### PR TITLE
skills: add basic methods for slot handling

### PR DESCRIPTION
**Pull request by zyga**
This branch adds five related functions:

Repository.AddSlot() adds a new slot to the repository.
Repository.RemoveSlot() removes a slot by snap and slot name.
Repository.AllSlots() returns all slots of a given type.
Repository.AllSlots() can also be used to look for any type, returning all slots in general.
Repository.Slots() returns a list of slots offered by a snap
Repository.Slot() looks up a skill by snap and slot name.
    
Internally there are also several helpers prefixed with the string
"unlocked" that aid in the implementation of the public methods.

### CODE REVIEW

**Comment by niemeyer** [comment]
This is almost a mirror of the skills API, so the same comments as the review of skills would apply here too.

**Comment by zyga** [comment]
Refreshed to match the fixes done in sklls-repo-skills.

File: skills/repo.go [hunk 1]
```diff
@@ -157,6 +159,93 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
...
+	if err := ValidateName(slot.Name); err != nil {
+		return err
+	}
+	// TODO: ensure that apps are correct
+	t := r.types[slot.Type]
+	if t == nil {
+		return fmt.Errorf("cannot add slot, skill type %q is not known", slot.Type)
+	}
+	if _, ok := r.slots[slot.Snap][slot.Name]; ok {
+		return fmt.Errorf("cannot add slot, snap %q already has slot %q", slot.Snap, slot.Name)
```
Comment (niemeyer) [inline skills/repo.go]: `"cannot add skill slot, ..."`

We should probably avoid speaking externally about slots without the context of skills.

**Comment by niemeyer** [comment]
LGTM, just a single trivial.

File: skills/repo.go [hunk 2]
```diff
@@ -157,6 +159,93 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
...
+	if r.slots[slot.Snap] == nil {
+		r.slots[slot.Snap] = make(map[string]*Slot)
+	}
+	r.slots[slot.Snap][slot.Name] = slot
+	return nil
+}
+
+// RemoveSlot removes a named slot from the given snap.
+// Removing a slot that doesn't exist returns an error.
+// Removing a slot that uses a skill returns an error.
```
Comment (chipaca) [inline skills/repo.go]: does it?

File: skills/repo.go [hunk 3]
```diff
@@ -157,6 +159,93 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
...
+func (r *Repository) RemoveSlot(snapName, slotName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// TODO: return an error if slot is occupied by at least one capability.
```
Comment (chipaca) [inline skills/repo.go]: ah, ok then

Comment (zyga) [inline skills/repo.go]: Yeah, this is implemented in the -grantrevoke branch.

**Pull request merged**